src/csharp/Grpc.Tools/GeneratorServices.cs

```diff
@@ -55,31 +54,58 @@ protected bool GrpcOutputPossible(ITaskItem proto)
                 && !gsm.EqualNoCase("false");
         }
 
-        public abstract string[] GetPossibleOutputs(ITaskItem proto);
+        // Update OutputDir and GrpcOutputDir for the item and all subsequent
+        // targets using this item. This should only be done if the real
+        // output directories for protoc should be modified.
+        public virtual void PatchOutputDirectory(ITaskItem protoItem)
+        {
+            // Nothing to do
+        }
+
+        public abstract string[] GetPossibleOutputs(ITaskItem protoItem);
     };
 
     // C# generator services.
     internal class CSharpGeneratorServices : GeneratorServices
     {
         public CSharpGeneratorServices(TaskLoggingHelper log) : base(log) { }
 
+        public override void PatchOutputDirectory(ITaskItem protoItem)
+        {
+            string root = protoItem.GetMetadata(Metadata.ProtoRoot);
+            string proto = protoItem.ItemSpec;
+            string relative = DepFileUtil.GetRelativeDir(root, proto, Log);
+
+            string outdir = protoItem.GetMetadata(Metadata.OutputDir);
+            string pathStem = Path.Combine(outdir, relative);
+            protoItem.SetMetadata(Metadata.OutputDir, pathStem);
+
+            // Override outdir if GrpcOutputDir present, default to proto output.
+            string grpcdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
+            if (grpcdir != "")
+            {
+                pathStem = Path.Combine(grpcdir, relative);
+            }
+            protoItem.SetMetadata(Metadata.GrpcOutputDir, pathStem);
```

Comment:
I feel that updating the GrpcOutputDir and OutputDir makes the logic quite hard to follow?
Who will be the changes visible to?  Does this actually modify the input item (visible to the entire rest of the build) or is this local to this task (in which case we should probably remember the values in a different way?)

Reply:
This will be used to update `Protobuf_Compile` for the rest of the build and it's documented in the `targets` file to make it easier to follow. This allows us to streamline the build. We could also add a separate metadata instead of overwriting it. But then, the original `OutputDir` would not be used anywhere in the `targets`.

For C++, this is skipped as protoc is generating the sub-folders according to this logic already. For C#, we would adjust the `OutputDir` to point to the relative path as C++ and other languages would do it.

```
Identity: protos/sub/folder/foo.proto
ProtoRoot: protos
OutputDir (before patch): obj/Debug/netstandard2.1
OutputDir (after patch): obj/Debug/netstandard2.1/sub/folder
```

Therefore the plugin will generate the files to `obj/Debug/netstandard2.1/sub/folder` as if the user had manually set `OutputDir = "$(Protobuf_OutputPath)%(RelativeDir)"` but this also works for protos not on the drive since `ProtoRoot` is used and not the project location. This just replicates the C++ behavior.